Code/Core/Strings/AString.h

```diff
@@ -92,6 +92,8 @@ class AString
     uint32_t Replace( const char * from, const char * to, uint32_t maxReplaces = 0 );
     void ToLower();
     void ToUpper();
+    void JSONEscape();
+    void Unquote();
```

Comment:
Are these functions generic enough to be in the `AString` header, or should I move them to `CompilationDatabase` class?

Reply:
As they only have one user, I think moving them into the compilation database would make sense.
That would also allow the JSONEscape to be greatly simplified, since it would no longer need to handle most of what it currently does, because we know the use is contstrained.

Reply:
OK, I will move them.
But I don't think that `JSONEscape` can be simplified much, because:
1. It still has to be able to escape all characters that need escaping (they can appear inside defines in `.CompilerOptions`), so the giant switch won't get any smaller.
2. It is still beneficial to calculate `extraSpace` before modifying anything because most of the time there will be no characters to escape. Here are some statistics for the FASTBuild repo:
On Linux running `fbuild -compdb All-x64Linux-Debug` results in 615 calls to `JSONEscape` and all of them are no-op (exit via `extraSpace == 0` condition).
On Windows running `FBuild -compdb All-x64-Debug` results in 794 calls to `JSONEscape`, 371 of them are no-op and only 10 call `Grow`. And if `/` would have been used as a path separator in Windows parts of bff files then there would be even more no-op cases.

One thing that can be changed is the direction of the second loop. This, I guess, will make it simpler but it will also pessimize cases in which there are enough reserved space to hold the modified string (which, as the stats above show, is the majority of cases).
So, should I change the loop direction?